src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,147 @@
+        private Node<T> next;
+
+        public Node(Node<T> prev, T element, Node<T> next) {
+            this.prev = prev;
+            this.element = element;
+            this.next = next;
+            size++;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (head == null) {
+            Node<T> firstNode = new Node<>(null, value, null);
+            head = firstNode;
+            tail = firstNode;
+        } else {
+            Node<T> nextNode = new Node<>(tail, value, null);
+            tail.next = nextNode;
+            tail = nextNode;
+        }
     }
 
     @Override
     public void add(T value, int index) {
+        indexIsValidAdd(index);
+        if (index == size()) {
+            add(value);
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head.prev = newNode;
+            head = newNode;
+        } else {
+            addToTheMiddle(value, index);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T node : list) {
+            add(node);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> node = getNode(index);
+        T oldValue = node.element;
+        node.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        return unlink(getNode(index));
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size(); i++) {
+            T currentElement = getNode(i).element;
+            if ((currentElement == object)
+                    || (currentElement != null && currentElement.equals(object))) {
+                remove(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private boolean indexIsValid(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException(EXCEPTION_MESSAGE);
+        }
+        return true;
+    }
+
+    private void addToTheMiddle(T value, int index) {
+        Node<T> current = head;
+        int count = 0;
+        while (count++ != index) {
+            current = current.next;
+        }
+        Node<T> newNode = new Node<>(current.prev, value, current);
+        newNode.prev.next = newNode;
+        newNode.next.prev = newNode;
+    }
+
+    private Node<T> getNode(int index) {
+        indexIsValid(index);
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private T unlink(Node<T> someNode) {
```

Comment:
Same. No reasone to make lock in separate method

Reply:
We should have `unlink` as separate method: Checklist: Create a separate method unlink(Node node).
This logic will be used in 2 methods: when we remove from list by index and by value. So let’s create a method that will take a Node that needs to be removed and unlink it. As a result we will call it from both methods when Node is found.